### PR TITLE
lexer, parser: Decorate errors with expression snippet.

### PR DESCRIPTION
**Pull request by qinqon**
At lexer, parser is easy to detect the rune, token that is generating
the error. This change wrap those errors with the expression and a
pointer.
    
Since the parser knows only about tokens, this change add also the
expression to the parser so it can use it to wrap the errors.

At resolver the ast.Node contains a Position field that points to the expressions
rune where the ast.Node was generated from. This change use that
position to to wrap the errors with a snippet that contains the
expression and a pointer to the place that generated the error.

To acomplish that the capture expressions are passed to the resolver so
it can search for the expression string to create the snippet.

The snippet pointer marks until operators but not inside the paths, this
can be fixed at a follow up.

The errors will look like:
```
invalid equality filter: right hand argument is not a string
| routes.running.destination==this.is.wrong                                     
| ............................^
```

### CODE REVIEW

**Comment by qinqon** [comment]
/hold
I am going to add some error to integration test to see how it looks

**Comment by qinqon** [comment]
/hold cancel

Integration tests for lexer/parser/resolver error has being added.

File: nmpolicy/internal/resolver/resolver.go [hunk 1]
```diff
@@ -80,34 +94,37 @@ func (r *resolver) resolveCaptureEntryName(captureEntryName string) (types.NMSta
...
-	} else if captureASTEntry.Replace != nil {
-		return r.resolveReplace(captureASTEntry.Replace)
+func (r resolver) resolveCaptureASTEntry() (types.NMState, error) {
+	if r.currentNode.EqFilter != nil {
+		return r.resolveEqFilter()
+	} else if r.currentNode.Replace != nil {
+		return r.resolveReplace()
 	}
-	return nil, fmt.Errorf("root node has unsupported operation : %+v", captureASTEntry)
+	return nil, fmt.Errorf("root node has unsupported operation : %+v", *r.currentNode)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver.go:114]: Let's not print `%+v` it is not readable for users. Let's add `String()` method to `ast.Node`. We can address this in a follow-up.

File: nmpolicy/internal/resolver/resolver.go [hunk 2]
```diff
@@ -155,14 +175,14 @@ func (r resolver) resolveInputSource(inputSourceNode ast.Node) (types.NMState, e
 		return capturedState, nil
 	}
 
-	return nil, fmt.Errorf("invalid input source %v, only current state or capture reference is supported", inputSourceNode)
+	return nil, fmt.Errorf("invalid input source %v, only current state or capture reference is supported", *r.currentNode)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver.go:178]: same here

File: nmpolicy/internal/resolver/resolver_test.go [hunk 3]
```diff
@@ -370,13 +372,16 @@ base-iface-routes:
...
-		runTest(t, testToRun)
+		runTest(t, &testToRun)
 	})
 }
 
 func testFilterInvalidTypeOnPath(t *testing.T) {
 	t.Run("Filter invalid type on path", func(t *testing.T) {
 		testToRun := test{
+			captureExpressions: `
+invalid-path-type: "01234567"
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver_test.go:383]: Can you please use the real capture expression `invalid-path-type: "interfaces.ipv4.address==10.244.0.1"` instead of `"01234567"`?
For me it was hard to follow and see it the `^` place makes sense when using `"01234567"`.

File: tests/basic_test.go [hunk 4]
```diff
@@ -439,6 +443,79 @@ func testGenerateUniqueTimestamps(t *testing.T) {
...
+		}
+		_, err := nmpolicy.GenerateState(policySpec, stateData, types.CachedState{})
+		assert.EqualError(t, err, `failed to generate state, err: failed to resolve capture expression, err: illegal rune -
+| routes.running.destination==-"0.0.0.0/0"
+| ............................^`)
+	})
+}
+
+func testFailureParser(t *testing.T) {
+	t.Run("with lexer error", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:470]: with parser error

File: tests/basic_test.go [hunk 5]
```diff
@@ -439,6 +443,79 @@ func testGenerateUniqueTimestamps(t *testing.T) {
...
+		assert.EqualError(t, err,
+			"failed to generate state, err: failed to resolve capture expression, "+
+				"err: invalid pipe: only paths can be piped in"+`
+| routes.running.destination=="0.0.0.0/0" |
+| ........................................^`)
+	})
+}
+
+func testFailureResolver(t *testing.T) {
+	t.Run("with lexer error", func(t *testing.T) {
```
Comment (AlonaKaplan) [inline tests/basic_test.go:495]: with resolver error

File: tests/basic_test.go [hunk 6]
```diff
@@ -439,6 +443,79 @@ func testGenerateUniqueTimestamps(t *testing.T) {
...
+				"cap1": `interfaces | routes.running.destination=="0.0.0.0/0"`,
+			},
+			DesiredState: stateData,
+		}
+		_, err := nmpolicy.GenerateState(policySpec, stateData, types.CachedState{})
+		assert.EqualError(t, err,
+			"failed to generate state, err: failed to resolve capture expression, "+
+				"err: resolve error: eqfilter error: invalid path input source, only capture reference is supported"+`
+| interfaces | routes.running.destination=="0.0.0.0/0"
+| .......................................^`)
```
Comment (AlonaKaplan) [inline tests/basic_test.go:515]: in this specific case it would be more readable if the `^` would point to the `interfaces` or to the pipe. We can fix it in a follow-up.

File: nmpolicy/internal/resolver/resolver.go (hunk 1 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver.go:114]: This should be addressed as part of fix on #44

File: nmpolicy/internal/resolver/resolver.go (hunk 2 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver.go:178]: ditto

File: nmpolicy/internal/resolver/resolver_test.go (hunk 3 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:383]: Problem with that is that make more difficult to put correct pos at AST.

File: tests/basic_test.go (hunk 6 above)
Comment (qinqon) [reply tests/basic_test.go:515]: We will fix it at follow up after merging the PR that use a visitor for path.go

File: nmpolicy/internal/resolver/resolver_test.go (hunk 3 above)
Comment (AlonaKaplan) [reply nmpolicy/internal/resolver/resolver_test.go:383]: I understand, but I believe it worth the effort since it will make the test much more readable.

File: tests/basic_test.go (hunk 6 above)
Comment (qinqon) [reply tests/basic_test.go:515]: Added the visitor part at the issue that covers the path.go refactoring https://github.com/nmstate/nmpolicy/issues/50.

**Comment by qinqon** [comment]
I have mix a wip on different PR here that's why integration tests is failing...

**Comment by qinqon** [comment]
@AlonaKaplan PR ready now, have removed the noise from other PR.

File: nmpolicy/internal/resolver/resolver_test.go [hunk 7]
```diff
@@ -73,17 +77,38 @@ interfaces:
...
+	for captureEntryName, captureEntryExpression := range typestest.ToCaptureExpressions(t, captureExpressions) {
+		l := lexer.New()
+		tokens, err := l.Lex(captureEntryExpression)
+		assert.NoError(t, err)
+		p := parser.New()
+		astRoot, err := p.Parse(captureEntryExpression, tokens)
+		assert.NoError(t, err)
+		captureASTPool[captureEntryName] = astRoot
+	}
+	captureASTPoolMarshaled, err := yaml.Marshal(captureASTPool)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver_test.go:99]: Why marshalling the pool is required if we unmarshal it anyway in the `runTest`?

Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:99]: So we don't have to change the rest of the code, since it's a test is not that problematic.

Comment (AlonaKaplan) [reply nmpolicy/internal/resolver/resolver_test.go:99]: What rest of the code?
Do you just need to change in `runTest`-
`captureASTPool := typestest.ToCaptureASTPool(t, testToRun.captureASTPool)` to
`captureASTPool := testToRun.captureASTPool`

Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:99]: then we have to change all the test that are using "captureASTPool" as a string.

**Review by AlonaKaplan — APPROVED** [review]
